components/omega/src/base/Halo.cpp

```diff
@@ -664,23 +696,47 @@ int Halo::startReceives() {
 // Initiate MPI communication by calling MPI_Isend for each Neighbor to send
 // the packed buffers to each task
 
-int Halo::startSends() {
+int Halo::startSends(const bool UseDevBuffer) {
 
    // Initialize vector to track MPI errors for each MPI_Isend
    std::vector<I4> IErr(NNghbr, 0);
 
    I4 Err{0}; // Error code to return
 
+   if (UseDevBuffer)
+      Kokkos::fence();
+
    for (int INghbr = 0; INghbr < NNghbr; ++INghbr) {
-      if (SendFlags[MyElem][INghbr]) {
-         MyNeighbor    = &Neighbors[INghbr];
-         I4 BufferSize = TotSize * MyNeighbor->SendLists[MyElem].NTot;
+      if (SendFlags[CurElem][INghbr]) {
+         auto &LocNeighbor = Neighbors[INghbr];
+
+         void *DataPtr{nullptr};
+
+         I4 BufferSize = TotSize * LocNeighbor.SendLists[CurElem].NTot;
+
+         // If UseDevBuffer is true the device buffer was packed by packBuffer,
+         // otherwise the host buffer was
+         if (UseDevBuffer) {
+            // If ExchOnDev is true, the device buffer can be passed to
+            // MPI_Isend, otherwise the device buffer needs to be copied
+            // to the host buffer, which will be passed to MPI_Isend
+            if (ExchOnDev) {
+               DataPtr = LocNeighbor.SendBuffer.data();
+            } else {
+               Kokkos::resize(LocNeighbor.SendBufferH, BufferSize);
+               deepCopy(LocNeighbor.SendBufferH, LocNeighbor.SendBuffer);
+               DataPtr = LocNeighbor.SendBufferH.data();
+            }
```

Comment:
When `OMEGA_ARCH=SERIAL` and `OMEGA_MPI_ON_DEVICE=OFF`, `UseDevBuffer` is `true` and `ExchOnDev` is `false`. Therefore, unnecessary `deepCopy` is executed in the `else` block.

Reply:
For CPU-only builds like `OMEGA_ARCH` is `SERIAL` or `OPENMP` there shouldn't be any need to use `OMEGA_MPI_ON_DEVICE=OFF`, part of the reason why what I designed calls for it to be on by default. `OMEGA_MPI_ON_DEVICE` primarily determines if the device or host buffer is passed to MPI calls, and if there is no GPU we don't need to worry about GPU-aware MPI support.